### PR TITLE
[WIP] Refactor the webui integration test to support external image

### PR DESCRIPTION
**Pull request by tmckayus**
This change does an initial refactor of the integration test so that it can be
run using a local image or an external image. It adds a test/e2e.sh
script for running the test which handles the details of pushing
images to integrated/external registries or allowing the image
to be referenced directly from the host in the case of 'oc cluster up'.

### CODE REVIEW

File: test/e2e.sh [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+# Modify the template if we're using a local image with no registry, i.e. we're in an oc cluster up case
+# In this case we don't need a push at all, but we to have a pullpolicy of IfNotPresent
+tweak_template
+print_test_env
+# Push the image if required
+push_image
+
+set +e
+oc create -f $WEBUI_TEST_RESOURCES
+oc policy add-role-to-user admin system:serviceaccount:$PROJECT:oshinko
```
Comment (tmckayus) [inline test/e2e.sh]: this line here actually shouldn't be necessary, the rolebinding in the template out to take care of it. Will remove in the next push.

Comment (crobby) [reply test/e2e.sh]: Sure, if it works without it, fantastic.  For some reason, I thought that just "edit" in the project wasn't enough.  If the push fails, maybe the rolebinding needs to be tweaked.

File: test/e2e.sh [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+	fi
+    fi
+    if [ -n "$registry" ]; then
+	set +e
+	docker login --help | grep email &> /dev/null
+	res=$?
+	set -e
+	if [ -n "$password" ]; then
+	    if [ "$res" -eq 0 ]; then
+		docker login -u ${user} -e jack@jack.com -p ${password} ${registry}
```
Comment (crobby) [inline test/e2e.sh]: should be able to remove the -e <email>

**Comment by crobby** [comment]
Another thing that might need tweaking is to run oc cluster up with a 'real' (not 127.0.0.1) IP address.  In the past, I think some of the SSL bits become challenged (oauth proxy/oc-proxy) when using 127.0.0.1.  I suppose ymmv though.

File: test/e2e.sh (hunk 2 above)
Comment (tmckayus) [reply test/e2e.sh]: I had this issue somewhere, different versions of docker. Maybe that's past now.

**Comment by tmckayus** [comment]
ack, if creation of the cluster is left external to e2e.sh then this shouldn't be a problem at all.  It will use whatever route is there.  Still migrating bits from travis.yaml ...

**Comment by crobby** [comment]
ah, gotcha.  I was thinking oc cluster up was going to find its way into here as well.  Great.

**Comment by crobby** [comment]
Should probably tweak the README.md with info on how to run this

File: test/e2e.sh [hunk 3]
```diff
@@ -0,0 +1,227 @@
...
+if [ -z "$WEBUI_TEST_IMAGE" ]; then
+    if [ "$WEBUI_TEST_LOCAL_IMAGE" == true ]; then
+        WEBUI_TEST_IMAGE=oshinko-webui:latest
+    else
+        WEBUI_TEST_IMAGE=docker.io/radanalyticsio/oshinko-webui
+    fi
+fi
+
+function print_test_env {
+    echo Using image $WEBUI_TEST_IMAGE
```
Comment (tmckayus) [inline test/e2e.sh]: hmm, this thing should print whether it's a secure test or not and whether or not the user/pass has been overridden

File: test/e2e.sh [hunk 4]
```diff
@@ -0,0 +1,227 @@
...
+}
+
+function push_image {
+    # The ip address of an internal/external registry may be set to support running against
+    # an openshift that is not "oc cluster up" when using images that have been built locally.
+    # In the case of "oc cluster up", the docker on the host is available from openshift so
+    # no special pushes of images have to be done.
+    # In the case of a "normal" openshift cluster, a local image we'll use for build has to be
+    # available from the designated registry.
+    # If we're using a pyspark image already in an external registry, openshift can pull it from
```
Comment (tmckayus) [inline test/e2e.sh]: typo, pyspar, can you tell where I cribbed this from? :)

File: test/e2e.sh [hunk 5]
```diff
@@ -0,0 +1,227 @@
...
+
+function tweak_template {
+    if [ "$WEBUI_TEST_LOCAL_IMAGE" == true ] && [ -z "$WEBUI_TEST_INTEGRATED_REGISTRY" ] && [ -z "$WEBUI_TEST_EXTERNAL_REGISTRY" ]; then
+	echo Updating $WEBUI_TEST_RESOURCES to pull local webui image
+        if [ -f "$WEBUI_TEST_RESOURCES" ]; then
+            cp "$WEBUI_TEST_RESOURCES" `pwd`/resources_mod.yaml
+        else
+            wget "$WEBUI_TEST_RESOURCES" -O `pwd`/resources_mod.yaml	     
+        fi
+        # This sed command finds the line with image: ${OSHINKO_WEB_IMAGE}
```
Comment (tmckayus) [inline test/e2e.sh]: should say "any lines"

File: test/e2e.sh [hunk 6]
```diff
@@ -0,0 +1,227 @@
...
+            Xvfb -ac :99 -screen 0 1280x1024x16  &
+        else
+            echo xvfb is already running
+        fi
+        export DISPLAY=:99
+    fi
+}
+
+# Modify the template if we're using a local image with no registry, i.e. we're in an oc cluster up case
+# In this case we don't need a push at all, but we to have a pullpolicy of IfNotPresent
```
Comment (tmckayus) [inline test/e2e.sh]: typo, missing "need"

**Comment by tmckayus** [comment]
ack on a README.md for this ...

**Comment by crobby** [comment]
While the hood is open on this.  Should the Makefile be updated to include targets for testing?

File: test/e2e.sh (hunk 2 above)
Comment (tmckayus) [reply test/e2e.sh]: yeah, this is necessary for origin 1.5, different version of docker. I'm on the fence about leaving it, the s2i tests have it.

**Pull request merged**